ext/hook/uhook_legacy.c

```diff
@@ -0,0 +1,448 @@
+    }
+
+    if (def->begin) {
+        dyn->dropped_span = !dd_uhook_call(def->begin, def->tracing, dyn, execute_data, &EG(uninitialized_zval));
+        if (def->tracing && dyn->dropped_span) {
+            ddtrace_clear_execute_data_span(invocation, false);
+        }
+    }
+
+    return true;
+}
+
+// create an own span for every generator resumption
+static void dd_uhook_generator_resumption(zend_ulong invocation, zend_execute_data *execute_data, zval *value, void *auxiliary, void *dynamic) {
+    (void)value;
+
+    dd_uhook_def *def = auxiliary;
+    dd_uhook_dynamic *dyn = dynamic;
+
+    if (dyn->skipped || !dyn->was_primed) {
+        dyn->was_primed = true;
+        return;
+    }
+
+    if (!get_DD_TRACE_ENABLED()) {
+        dyn->dropped_span = true;
+        return;
+    }
+
+    if (def->tracing) {
+        dyn->span = ddtrace_alloc_execute_data_span(invocation, execute_data);
+        dyn->dropped_span = false;
+    }
+
+    if (def->begin) {
+        dyn->dropped_span = !dd_uhook_call(def->begin, def->tracing, dyn, execute_data, value);
+        if (def->tracing && dyn->dropped_span) {
+            ddtrace_clear_execute_data_span(invocation, false);
+        }
+    }
+}
+
+static void dd_uhook_generator_yield(zend_ulong invocation, zend_execute_data *execute_data, zval *key, zval *value, void *auxiliary, void *dynamic) {
+    (void)key;
+
+    dd_uhook_def *def = auxiliary;
+    dd_uhook_dynamic *dyn = dynamic;
+
+    if (dyn->skipped) {
+        return;
+    }
+
+    if (def->tracing && !dyn->dropped_span) {
+        if (dyn->span->span.duration == -1ull) {
+            dyn->dropped_span = true;
+            ddtrace_clear_execute_data_span(invocation, false);
+
+            if (get_DD_TRACE_ENABLED()) {
+                ddtrace_log_errf("Cannot run tracing closure for %s(); spans out of sync", ZSTR_VAL(EX(func)->common.function_name));
+            }
+        } else {
+            zval *exception_zv = ddtrace_spandata_property_exception(&dyn->span->span);
+            if (EG(exception) && Z_TYPE_P(exception_zv) <= IS_FALSE) {
+                ZVAL_OBJ_COPY(exception_zv, EG(exception));
+            }
+
+            dd_trace_stop_span_time(&dyn->span->span);
+        }
+    }
+
+    if (def->end && (!def->tracing || !dyn->dropped_span)) {
+        bool keep_span = dd_uhook_call(def->end, def->tracing, dyn, execute_data, value);
+        if (def->tracing && !dyn->dropped_span) {
+            ddtrace_clear_execute_data_span(invocation, keep_span);
+        }
+        dyn->dropped_span = true;
+    }
+}
+
+static void dd_uhook_end(zend_ulong invocation, zend_execute_data *execute_data, zval *retval, void *auxiliary, void *dynamic) {
+    dd_uhook_def *def = auxiliary;
+    dd_uhook_dynamic *dyn = dynamic;
+    bool keep_span = true;
+
+    if (dyn->skipped) {
+        return;
+    }
+
+    if (def->tracing && !dyn->dropped_span) {
+        if (dyn->span->span.duration == -1ull) {
```

Comment:
From what I've gathered, duration is set to `-1` only for dropped spans. Is that right? If so can we use a constant or define or something, or at least leave a comment.